service/storage_service.cc

```diff
@@ -2457,6 +2454,50 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
     for (auto& x : loaded_peer_features) {
         slogger.info("peer={}, supported_features={}", x.first, x.second);
     }
+
+    if (!_raft_experimental_topology) {
+        slogger.info("Booting in legacy operations mode because experimental raft topology is not enabled");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (utils::get_local_injector().is_enabled("force_gossip_based_join") && !_sys_ks.local().bootstrap_complete()) {
+        slogger.info("Booting in legacy topology operations mode because it was requested by an error injection");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (_group0->client().in_recovery()) {
+        slogger.info("Raft recovery - starting in legacy topology operations mode");
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (_group0->joined_group0()) {
+        slogger.info("Restarting in raft mode");
+        _topology_change_kind_enabled = topology_change_kind::raft;
```

Comment:
But can't it be "schema over raft" only mode?

Reply:
Hm, right. At this point of this PR (i.e. right after this commit) it is true that this piece of code is incorrect. It is later fixed by `storage_service: set the topology kind value based on group 0 state`. I admit I got a bit lost when trying to organize the PR into commits and order them.

I'll try to reorganize the commits a bit and/or fix this piece of code so that it makes more sense.

Reply:
Unless something substantial is broken by it at this point that prevents test from running no need to reorder (may be just fix the comment and in the later patches re-fix it back).